tests/e2e/privateregistry/privateregistry_test.go

```diff
@@ -0,0 +1,162 @@
+				g.Expect(err).NotTo(HaveOccurred())
+				for _, pod := range pods {
+					if strings.Contains(pod.Name, "helm-install") {
+						g.Expect(pod.Status).Should(Equal("Completed"), pod.Name)
+					} else {
+						g.Expect(pod.Status).Should(Equal("Running"), pod.Name)
+					}
+				}
+			}, "620s", "5s").Should(Succeed())
+			_, _ = e2e.ParsePods(kubeConfigFile, true)
+		})
+
+		It("Create new private registry", func() {
+			registry, err := e2e.RunCmdOnNode("sudo docker run -d -p 5000:5000 --restart=always --name registry registry:2 ", serverNodeNames[0])
+			fmt.Println(registry)
+			Expect(err).NotTo(HaveOccurred())
+
+		})
+		It("ensures registry is working", func() {
+			a, err := e2e.RunCmdOnNode("sudo docker ps -a | grep registry\n", serverNodeNames[0])
+			fmt.Println(a)
+			Expect(err).NotTo(HaveOccurred())
+
+		})
+		It("Should pull and image from dockerhub and send it to private registry", func() {
+			cmd := "sudo docker pull nginx"
+			_, err := e2e.RunCmdOnNode(cmd, serverNodeNames[0])
+			Expect(err).NotTo(HaveOccurred(), "failed: "+cmd)
+
+			nodeIP, err := e2e.FetchNodeExternalIP(serverNodeNames[0])
+			Expect(err).NotTo(HaveOccurred())
+
+			cmd = "sudo docker tag nginx " + nodeIP + ":5000/my-webpage"
+			_, err = e2e.RunCmdOnNode(cmd, serverNodeNames[0])
+			Expect(err).NotTo(HaveOccurred(), "failed: "+cmd)
+
+			cmd = "sudo docker push " + nodeIP + ":5000/my-webpage"
+			_, err = e2e.RunCmdOnNode(cmd, serverNodeNames[0])
+			Expect(err).NotTo(HaveOccurred(), "failed: "+cmd)
+
+			cmd = "sudo docker image remove nginx " + nodeIP + ":5000/my-webpage"
+			_, err = e2e.RunCmdOnNode(cmd, serverNodeNames[0])
+			Expect(err).NotTo(HaveOccurred(), "failed: "+cmd)
+		})
+		It("Should create and validate deployment with private registry on", func() {
+			res, err := e2e.RunCmdOnNode("sudo kubectl create deployment my-webpage --image=my-registry.local/my-webpage", serverNodeNames[0])
+			fmt.Println(res)
+			Expect(err).NotTo(HaveOccurred())
+
+			var pod e2e.Pod
+			Eventually(func(g Gomega) {
+				pods, err := e2e.ParsePods(kubeConfigFile, false)
+				for _, p := range pods {
+					if strings.Contains(p.Name, "my-webpage") {
+						pod = p
+					}
+				}
+				g.Expect(err).NotTo(HaveOccurred())
+				g.Expect(pod.Status).Should(Equal("Running"))
+				g.Expect(pod.Node).Should(Equal(agentNodeNames[0]))
+			}, "60s", "5s").Should(Succeed())
+
+			cmd := "curl " + pod.IP
+			Expect(e2e.RunCmdOnNode(cmd, serverNodeNames[0])).To(ContainSubstring("Welcome to nginx!"))
+		})
+
+	})
+})
+
+var failed bool
+var _ = AfterEach(func() {
+	failed = failed || CurrentSpecReport().Failed()
+})
+
+var _ = AfterSuite(func() {
+	r1, err := e2e.RunCmdOnNode("sudo docker rm -f registry", serverNodeNames[0])
+	fmt.Println(r1)
+	Expect(err).NotTo(HaveOccurred())
+	r2, err := e2e.RunCmdOnNode("sudo kubectl delete deployment my-webpage", serverNodeNames[0])
+	fmt.Println(r2)
+	Expect(err).NotTo(HaveOccurred())
```

Comment:
You should only run this if the test has not failed. Because if the test fails, we either want to
1) its local and the VMs are still active, allowing us to debug a failure, we don't want to do any cleanup
2) Its on CI and the VMs are getting deleted anyway, so it doesn't matter

Also don't print the results, if you want put the out only if an error
```
Expect(err).NotTo(HaveOccurred(), r1)
```

Reply:
If you want to add structured information to the report, instead of just printing it to stdout, you can use `ginkgo.AddReportEntry`. Here's an example of using this to add pod logs when a test has failed:
https://github.com/k3s-io/helm-controller/blob/master/test/framework/framework.go#L62-L82